packages/flutter_tools/lib/src/emulator.dart

```diff
@@ -0,0 +1,176 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:math' as math;
+
+import 'android/android_emulator.dart';
+import 'base/context.dart';
+import 'globals.dart';
+import 'ios/ios_emulators.dart';
+
+EmulatorManager get emulatorManager => context[EmulatorManager];
+
+/// A class to get all available emulators.
+class EmulatorManager {
+  /// Constructing EmulatorManager is cheap; they only do expensive work if some
+  /// of their methods are called.
+  EmulatorManager() {
+    // Register the known discoverers.
+    _emulatorDiscoverers.add(new AndroidEmulators());
+    _emulatorDiscoverers.add(new IOSEmulators());
+  }
+
+  final List<EmulatorDiscovery> _emulatorDiscoverers = <EmulatorDiscovery>[];
+
+  String _specifiedEmulatorId;
+
+  /// A user-specified emulator ID.
+  String get specifiedEmulatorId {
+    if (_specifiedEmulatorId == null || _specifiedEmulatorId == 'all')
+      return null;
+    return _specifiedEmulatorId;
+  }
+
+  set specifiedEmulatorId(String id) {
+    _specifiedEmulatorId = id;
+  }
+
+  /// True when the user has specified a single specific emulator.
+  bool get hasSpecifiedEmulatorId => specifiedEmulatorId != null;
+
+  /// True when the user has specified all emulators by setting
+  /// specifiedEmulatorId = 'all'.
+  bool get hasSpecifiedAllEmulators => _specifiedEmulatorId == 'all';
+
+  Stream<Emulator> getEmulatorsById(String emulatorId) async* {
+    final List<Emulator> emulators = await getAllAvailableEmulators().toList();
+    emulatorId = emulatorId.toLowerCase();
+    bool exactlyMatchesEmulatorId(Emulator emulator) =>
+        emulator.id?.toLowerCase() == emulatorId ||
+        emulator.name?.toLowerCase() == emulatorId;
+    bool startsWithEmulatorId(Emulator emulator) =>
+        emulator.id?.toLowerCase()?.startsWith(emulatorId) == true ||
+        emulator.name?.toLowerCase()?.startsWith(emulatorId) == true;
+
+    final Emulator exactMatch = emulators.firstWhere(
+        exactlyMatchesEmulatorId, orElse: () => null);
+    if (exactMatch != null) {
+      yield exactMatch;
```

Comment:
I tend not to use `yield`s myself, but it does look like we use it elsewhere in flutter_tools.

Reply:
Yeah, it felt a bit weird to me tbh (especially as we call one function, toList it, then filter again), was just how it was in the code I copied. I might rip it out though, I don't think there's any benefit to having an iterable here (we're `toList()`ing it further up anyway).

Reply:
Actually, turns out this is because we enumerator multiple discovered and yield the results (to avoid adding to a list). We don't need to do it in `getEmulatorsById` (which I've renamed to `getEmulatorsMatching` since it isn't strictly ID) however it might br weird if `getAllAvailableEmulators` and `getEmulatorsMatching` returned different types (Stream vs List).

So, if you have any opinions on this, I'm happy to change, but leaving as-is for now.

Reply:
It's not important to change for this PR, but you should be able to change the return type to `Future<List<Emulator>>`, return `[exactMatch]` above, and return a filtered view of the emulators below.

Reply:
Yeah, that's what this comment was about:

> however it might br weird if `getAllAvailableEmulators` and `getEmulatorsMatching` returned different types (`Stream` vs `List`).

Making `getEmulatorsMatching` return ` Future<List>` is simple, but I can't figure out how to change `getAllAvailableEmulators` to do the same to match:

```dart
  Stream<Emulator> getAllAvailableEmulators() async* {
    for (EmulatorDiscovery discoverer in _platformDiscoverers) {
      for (Emulator emulator in await discoverer.emulators) {
        yield emulator;
      }
    }
  }
```

I can do `await getAllAvailableEmulators().toList()` and that does it, but I can't figure out how to fix the method itself. It seems like there should be a simple way (without just putting it inside a nested function) to do that from within?

Reply:
something perhaps like:

```
Future<List<Emulator>> getAllAvailableEmulators() async {
  List<Emulator> emulators = <Emulator>[];
  for (EmulatorDiscovery discoverer in _platformDiscoverers) {
    for (Emulator emulator in await discoverer.emulators) {
        emulators.add(emulator);
    }
  }

return emulators;
}
```

It's simple; it's possible you could also use `Future.wait` for slightly more terse code.

Reply:
Heh, I thought there might be a cleaner way. I got a lint(?) on your code saying to use `.forEach`, which I had to swap for `Future.forEach` to get it to work; so I ended up with:

```dart
  /// Return the list of all available emulators.
  Future<List<Emulator>> getAllAvailableEmulators() async {
    final List<Emulator> emulators = <Emulator>[];
    Future.forEach(_platformDiscoverers, (EmulatorDiscovery discoverer) async {
      emulators.addAll(await discoverer.emulators);
    });
    return emulators;
  }
```

I'll find a way to eliminate that list without making the code unreadable one day!